### ISSUE TITLE
render_shortcode_for_preview when post_id is null

### ISSUE THREAD

**Issue by phh** [wp-shortcake/shortcake#661]
Hi there,

When using Gravityforms and adding a post element to the confirmation editor I'm getting a "Something's rotten in the state of Denmark" error.

My current fix is to filter into map_meta_cap and allow this if $post_id is null.

I guess there should either be a check if it's the WP editor in the edit view or allow the shortcode preview if no post_id is sent.

I've just updated to the latest version of shortcode-ui. The 0.6.2 version had the same issue.

Suggestions?

**Comment by goldenapples** [comment]
Yeah, I've run across that same issue before - see #599.

It's not going to be easy to anticipate all the possible conditions where a user might want to use shortcode previews in editors and define the right capability checks to handle them inside Shortcake. But maybe we could have a filter around the check for whether the user is allowed to preview, and let that be extended elsewhere.

Maybe if we localized the value of `current_screen()` and sent that along with every preview request, that could provide enough context for users to determine whether to allow previews in a given context? I'm thinking changing this simple check [here](https://github.com/wp-shortcake/shortcake/blob/master/inc/class-shortcode-ui.php#L363-L365) to something more like

```
$can_preview = current_user_can( 'edit_post', $post_id );

if ( false === apply_filters( 'shortcode_ui_allow_preview_capability', $can_preview, $current_screen_context ) ) {
    return esc_html__( "Something's rotten in the state of Denmark", 'shortcode-ui' );
}
```

would be enough to hook into.

Think that would work here?

**Comment by phh** [comment]
I think a filter would work. And having the current_screen in the filter it would give us alot of flexibility.

But I still think it's sad that the shortcode-ui plugin will not just work "out-of-the-box" with some plugins unless they add the capability check filter.

What is the exact reason for this capability check? It's not like the user is able to insert data from inside this ajax call as it just renders the HTML for the shortcode. Is check capability check really necessary?

**Comment by goldenapples** [comment]
> Is this capability check really necessary?

I think the idea here is just to mirror the capability required to preview a draft post, as a shortcode preview is essentially part of a draft post preview. A user who can't edit a draft post will get a 404 when trying to preview that draft; the capability check just makes that same behavior explicit for the shortcode preview.

One could easily argue that this isn't necessary, because to preview a shortcode you have to already have access to the shortcode's content. But there are a few cases where this isn't the case, like where a shortcode is a content block that's referenced by a post ID... a site might not want to expose previews of all of those content blocks to anyone who can craft the preview ajax POST request. Plus, since this is a draft preview, it makes sense to keep the same capability checks in place that core uses to allow access to view previews.

**Comment by phh** [comment]
Fair point. Would it not be better if shortcode-ui were only injected in the add/edit editor screens instead of in all editors? The `shortcode_ui_allow_preview_capability` would do the trick for now but this needs to be ticked for all screens.

The shortcode-ui could then be an option to add to your editor (and manage the capabilities yourself) instead of adding it be default and it not working out of the box.

**Comment by goldenapples** [comment]
> Would it not be better if shortcode-ui were only injected in the add/edit editor screens instead of in all editors?

That's probably a good idea; since the previews don't work on other screens anyways, it might be a good idea not to enqueue the scripts at all on those other pages unless specifically added by another filter - and then that filter would be responsible for also defining the capability required to render the preview.

I'm thinking this is a breaking change for some users who are using shortcake UI on term edit screens or options pages, so it'd require bumping the major version number. (Of course, I suppose we've been pretty loose with semantic versioning up to this point anyways - the 0.7.0 release should really have been 1.0.0 since it broke some backwards compatibility...)

**Comment by phh** [comment]
Maybe a new filter for adding shortcode-ui's to wp-editors with capability callback?

**Comment by entr** [comment]
At least don't do `current_user_can( 'edit_post', $post_id )` when `$post_id` is `null` since it has literally no way to pass.

**Comment by jameswburke** [comment]
Has anyone seen movement on this? I need to use shortcake on an options page. Happy to submit a PR.

**Comment by entr** [comment]
With the rise of Gutenberg coming soon I doubt if shortcake will receive much attention in the future.

**Comment by goldenapples** [comment]
> With the rise of Gutenberg coming soon I doubt if shortcake will receive much attention in the future.

Unfortunately, that seems to be true. 

However, this is a longstanding issue which would be good to fix. I've tried a couple times to build something out that intelligently determined the correct permissions check to use based on the admin screen and context, but wasn't ever satisfied with it. If you could put a pull request together, I'd be happy to review it.

**Comment by petenelson** [comment]
Ran into the same issue on a current project when adding a shortcode to a term edit page, and here's the code I dropped in to bypass it.

```
add_action( 'wp_ajax_bulk_do_shortcode', shortcode_ui_set_ajax_permissions', 5 );

/**
 * Sets up temporary permission filter for when trying to render the
 * email form shortcode via Shortcode UI for a term.
 *
 * @return void
 */
function shortcode_ui_set_ajax_permissions() {

	$post_id = null;

	if ( is_array( $_POST['queries'] ) ) {
		foreach ( $_POST['queries'] as $posted_query ) {
			if ( isset( $posted_query['post_id'] ) ) {
				$post_id = intval( $posted_query['post_id'] );
			} else {
				$post_id = null;
			}
		}
	}

	if ( empty( $post_id ) ) {
		// This lets us bypass the current_user_can() check in
		// render_shortcode_for_preview() in Shortcode UI.
		add_filter(
			'map_meta_cap',
			function() {
				if ( doing_action( 'wp_ajax_bulk_do_shortcode' ) ) {
					return [
						'edit_post',
					];
				}
			}
		);

		add_filter(
			'map_meta_cap',
			function() {
				if ( doing_action( 'wp_ajax_bulk_do_shortcode' ) ) {
					return [
						'edit_post',
					];
				}
			}
		);
	}
}
```
